### ISSUE TITLE
[bug] Adding mold linker to conanfile causes dependency loop

### ISSUE THREAD

**Issue by rdong8** [conan-io/conan#17333]
### Describe the bug

OS: Fedora 41
Compiler: Clang 19.1.0
Conan version: 2.9.2

### How to reproduce it

Add the sections from [this](https://github.com/conan-io/conan-center-index/tree/master/recipes/mold) page to your conanfile. Then run your `conan install`, ie.:

```
BUILD_DIR=build conan install . -b missing -pr:b local -pr:h default -s build_type=Release -s "&:build_type=Debug"     │
```

Result:

```
======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=26
compiler.libcxx=libc++
compiler.version=19
os=Linux
&:build_type=Debug
[platform_tool_requires]
cmake/3.31.0
[conf]
tools.cmake.cmaketoolchain:generator=Ninja Multi-Config
[buildenv]
CC=clang
CXX=clang++

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=26
compiler.libcxx=libc++
compiler.version=19
os=Linux
[tool_requires]
*: mold/[*]
[conf]
tools.build:exelinkflags=['-fuse-ld=mold']
tools.build:sharedlinkflags=['-fuse-ld=mold']


======== Computing dependency graph ========
Graph root
    conanfile.py: /home/rd8/Projects/cpp_project/conanfile.py
Requirements
    boost/1.86.0#46c984e8737c8b910597294db3838f3b - Cache
    bzip2/1.0.8#d00dac990f08d991998d624be81a9526 - Cache
    libbacktrace/cci.20210118#a7691bfccd8caaf66309df196790a5a1 - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Build requirements
    b2/5.2.1#91bc73931a0acb655947a81569ed8b80 - Cache
    mold/2.34.1#4d062e19dee8474c93eddf2af9ea3dfe - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Resolved version ranges
    b2/[>=5.2 <6]: b2/5.2.1
    mold/[*]: mold/2.34.1
    zlib/[>=1.2.11 <2]: zlib/1.3.1
ERROR: There is a cycle/loop in the graph:
    Initial ancestor: mold/2.34.1
    Require: mold/2.34.1
    Dependency: mold/2.34.1
```

**Comment by AbrilRBS** [comment]
Hi @rdong8 thanks a lot for taking the time to report this

This is probably not a bug, but expected behaviour.

Your `[tool_requires]` section on the `build` profile reads
```
[tool_requires]
*: mold/[*]
```

This means that every package in the `build` context will have `mold` injected as an extra `tool_require`.

As `*` also matches `mold`, this means that `mold` is getting itself injected as a `tool_require`

2 things to note here:
* You probably want this `[tool_require]` section on your `host` profile, not necessarily as much in the `build` one. As it stands right now, none of your actual libraries would be using `mold`.
* IF you also end up needed `mold` as part of your `build` context profile, you should specify the pattern as `!mold/*: mold/[*]` to exclude itself. Note that if `mold` has dependencies and those get the linker injected, you'll still have loops


Let me know if this helps :)

**Comment by rdong8** [comment]
Hey, thanks for the response. I have made the following changes:

build:

```
[buildenv]
CC=clang
CXX=clang++

[conf]
tools.cmake.cmaketoolchain:generator=Ninja Multi-Config

[platform_tool_requires]
cmake/3.31.0

[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=26
compiler.libcxx=libstdc++11
compiler.version=19
os=Linux
```

host:
```
[buildenv]
CC=clang
CXX=clang++

[conf]
tools.build:exelinkflags=['-fuse-ld=mold']
tools.build:sharedlinkflags=['-fuse-ld=mold']

[settings]
arch=x86_64
build_type=Release
compiler=clang
compiler.cppstd=26
compiler.libcxx=libc++
compiler.version=19
os=Linux

[tool_requires]
!mold/*: mold/[*]
```

However I still do have the issue with the dependency loop:

```
======== Computing dependency graph ========
Graph root
    conanfile.py: /home/rd8/Projects/cpp_project/conanfile.py
Requirements
    boost/1.86.0#46c984e8737c8b910597294db3838f3b - Cache
    bzip2/1.0.8#d00dac990f08d991998d624be81a9526 - Cache
    catch2/3.7.1#431d772165ed0bc5adaabaa44a9f53ca - Cache
    libbacktrace/cci.20210118#a7691bfccd8caaf66309df196790a5a1 - Cache
    spdlog/1.14.1#4fd40d9cbc1978247443a10d2ace58fd - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Build requirements
    b2/5.2.1#91bc73931a0acb655947a81569ed8b80 - Cache
    cmake/3.31.0 - Platform
    mold/2.34.1#4d062e19dee8474c93eddf2af9ea3dfe - Cache
    xxhash/0.8.2#03fd1c9a839b3f9cdf5ea9742c312187 - Cache
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2 - Cache
Resolved version ranges
    b2/[>=5.2 <6]: b2/5.2.1
    mold/[*]: mold/2.34.1
    zlib/[>=1.2.11 <2]: zlib/1.3.1

======== Computing necessary packages ========
spdlog/1.14.1: Main binary package '4c4d5237d406fd4c79c7bef05af533c8939e3c58' missing
spdlog/1.14.1: Checking 5 compatible configurations
spdlog/1.14.1: Compatible configurations not found in cache, checking servers
spdlog/1.14.1: '6679d20ab72b8fd423ecd46d0746e46c39770bcf': compiler.cppstd=20
spdlog/1.14.1: '60e040ab069a2b5d5c85ecd731af4cf86f9c230c': compiler.cppstd=gnu20
spdlog/1.14.1: '603d38981bae7e65f2e73ff7f9f51e59a201279e': compiler.cppstd=23
spdlog/1.14.1: '494d27793e5f9f33c2929a3b7c28a43c29aead3f': compiler.cppstd=gnu23
spdlog/1.14.1: '0b9f5ef239a0d52320e31bb08c49326c6a4621c3': compiler.cppstd=gnu26
boost/1.86.0: Main binary package 'c123612e482b0a8ecc61a6ddb0a62d25e7578366' missing
boost/1.86.0: Checking 5 compatible configurations
boost/1.86.0: Compatible configurations not found in cache, checking servers
boost/1.86.0: 'af3e0258503efbefd0af55b4f2db656adc861436': compiler.cppstd=20
boost/1.86.0: '3427da6b3ac26939436d8224d9e5ec48f192a412': compiler.cppstd=gnu20
boost/1.86.0: 'cbf236643def2491b4668a4a48b7a8131c6d5623': compiler.cppstd=23
boost/1.86.0: 'c9e387a2d958cc83509b4521866344929375614a': compiler.cppstd=gnu23
boost/1.86.0: '728d5f8f5ab862ab57258b1bfca76baba484a027': compiler.cppstd=gnu26
Requirements
    boost/1.86.0#46c984e8737c8b910597294db3838f3b:c123612e482b0a8ecc61a6ddb0a62d25e7578366 - Build
    bzip2/1.0.8#d00dac990f08d991998d624be81a9526:ab19b9bfc1b47ea17e1a675221a2f0b1e27ebba5 - Build
    catch2/3.7.1#431d772165ed0bc5adaabaa44a9f53ca:882cd1aca82e6fa218c2563f3d8e8bee421ac4a8 - Build
    libbacktrace/cci.20210118#a7691bfccd8caaf66309df196790a5a1:035983102d94f5d9d7256017a5fd89651d1294c3 - Build
    spdlog/1.14.1#4fd40d9cbc1978247443a10d2ace58fd:4c4d5237d406fd4c79c7bef05af533c8939e3c58 - Build
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:035983102d94f5d9d7256017a5fd89651d1294c3 - Build
Build requirements
    b2/5.2.1#91bc73931a0acb655947a81569ed8b80:63fead0844576fc02943e16909f08fcdddd6f44b#f9c46d924270b1920fa5398114be2576 - Download (conancenter)
    cmake/3.31.0 - Platform
    mold/2.34.1#4d062e19dee8474c93eddf2af9ea3dfe:77dc507434abadcc87e1517333a94365a4b99b07 - Build
    xxhash/0.8.2#03fd1c9a839b3f9cdf5ea9742c312187:aa3784fb750b4514cd847a028afd1f7efa4e7711 - Build
    zlib/1.3.1#f52e03ae3d251dec704634230cd806a2:035983102d94f5d9d7256017a5fd89651d1294c3 - Build
ERROR: There is a loop in the graph (some packages already ommitted):
zlib/1.3.1 (Build) -> ['mold/2.34.1']
mold/2.34.1 (Build) -> ['zlib/1.3.1', 'xxhash/0.8.2', 'cmake/3.31.0']
bzip2/1.0.8 (Build) -> ['mold/2.34.1']
catch2/3.7.1 (Build) -> ['mold/2.34.1']
libbacktrace/cci.20210118 (Build) -> ['mold/2.34.1']
spdlog/1.14.1 (Build) -> ['mold/2.34.1']
boost/1.86.0 (Build) -> ['zlib/1.3.1', 'bzip2/1.0.8', 'libbacktrace/cci.20210118', 'b2/5.2.1', 'mold/2.34.1']

This graph is ill-formed, and cannot be installed
Most common cause is having dependencies both in build and host contexts
forming a cycle, due to tool_requires having transitive requires.
Check your profile [tool_requires] and recipe tool and regular requires
You might inspect the dependency graph with:
  $ conan graph info . --format=html > graph.html
```

Would it help if I installed mold external to conan and somehow just told conan to use that? I'm not sure if I follow fully but it seems like part of the problem is mold itself needing to be built.

**Comment by perseoGI** [comment]
Greetings @rdong8,

This is a known limitation with Conan.

**TL;DR**: First, compile `mold` separately from your project using the following command:
```sh
$ conan install --requires="mold/[*]" --build=missing
```
Once you've done that, you should be able to compile your project without issues.

**Explanation**: 
The `mold` tool depends on `zlib`, `xxhash`, and optionally on `mimalloc`. If your project also depends on any of these libraries, specifically `zlib`, and you haven't already compiled `mold`, a conflict will occur. When Conan builds the dependency graph, there will be a conflict between the host dependency (`zlib`) and the `zlib` dependency required by `mold` in the build context. Because of this, Conan will raise a cycle error since it cannot distinguish between the two dependencies in different contexts.

This behavior is fundamental and not typical, so there isn't a straightforward solution. The best approach is to compile the tool (`mold`) first, and then proceed with your project.

I hope this clarifies things! Please let me know if following these instructions resolves your issue.

**Comment by memsharded** [comment]
I'd like to understand this a bit better, and consider other possible approaches, maybe the ``vendor=True``? 

I have just checked that we don't have a test that checks the ``This graph is ill-formed, and cannot be installed`` error message, I'd like to double check that some of our tests is capturing this scenario too.

**Comment by rdong8** [comment]
Coming back to this:

>You probably want this [tool_require] section on your host profile, not necessarily as much in the build one. As it stands right now, none of your actual libraries would be using mold.

Sorry not fully following (not familiar with cross-compilation in general). Don't we need the linker in the build profile along with the compiler, CMake, etc? For instance from [this article](https://docs.conan.io/2/tutorial/consuming_packages/cross_building_with_conan.html):

>The build context contains the tool requirements used on the build machine. This category typically includes all the developer tools like CMake, compilers and linkers. In this case, this includes the cmake/3.27.9 tool.

**Comment by perseoGI** [comment]
Greetings @rdong8 

I think what’s confusing here is mixing up host vs build profiles in Conan. They serve different roles:

#### Host vs Build Profile
* The build profile is often misunderstood. Some users think it defines the tools needed to build your project, but in reality it defines the toolchain, settings, options, and dependencies that are used to build everything that runs in the build machine context.
* The host profile defines the settings, options, and dependencies for the binaries that will run in the host machine context (i.e., your target system).

##### Example: If your recipe `requires=zlib` and `tool_requires=cmake`:
* zlib and your final app/library will be built using the host profile.
* cmake will be built (or fetched) using the build profile because it runs during the build process.


#### Common Misunderstandings
* Putting `tool_requires` in the build profile means that the specified tool (e.g. `mold`, `cmake`) is used for building other tools or build-utilities, but not for linking `zlib` or the final library. If you want `mold` or a special linker for your app/library, that should live in the host profile instead.
* Similarly, linkers, compilers for code generation for the target should generally be host settings, unless you are explicitly building tools or cross-compilers.

#### Cross-Compilation & buildenv

When cross compiling:
* The host profile should set the target OS/architecture (e.g. Android armv8, iOS arm64, etc.).
* If you use a cross-compiler (or a toolchain package), that should normally be placed in the `[buildenv]` section of the host profile. This way Conan generates the proper environment so that tools used during the build (but running on your build machine) use the correct cross-compiler binaries, sysroot, etc.


So answering your question, `mold` and any kind of linkers you want to use to compile your application should be in the host profile.

**Comment by rdong8** [comment]
Thank you for the explanation! It seems like this was indeed the issue. Since I didn't actually need cross compilation I also ended up getting it to work by reducing to a single profile and setting `CMAKE_LINKER_TYPE` directly since I already had mold installed:

```ini
[conf]
tools.cmake.cmaketoolchain:extra_variables={'CMAKE_LINKER_TYPE': 'MOLD'}
```

**Comment by perseoGI** [comment]
I'm glad you could manage to address the issue.
If you have already installed the linker, following your approach is a good idea. This way, only having a single profile, and if you use that profile for host and build, you could also link your build dependencies with mold which is great.

**Comment by tttapa** [comment]
@perseoGI Could this issue please be reopened?

As far as I can tell, the fundamental problem has not been resolved.  
Minimal reproducible example:

```ini
# host.profile
include(default)
[tool_requires]
mold/2.32.1
```
```sh
conan install --require zlib/1.3.1 -pr:h host.profile --build=missing
```

<details><summary>Output ...</summary>

```
======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=14
os=Linux
[tool_requires]
*: mold/2.32.1

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=14
os=Linux


======== Computing dependency graph ========
Graph root
    cli
Requirements
    zlib/1.3.1#b8bc2603263cf7eccbd6e17e66b0ed76 - Cache
Build requirements
    cmake/3.31.6#ed0e6c1d49bd564ce6fed1a19653b86d - Cache
    mold/2.32.1#e31d60209daa5ece4edf21dd3e7c15f5 - Cache
    xxhash/0.8.2#7856c968c985b2981b707ee8f2413b2b - Cache
    zlib/1.3.1#b8bc2603263cf7eccbd6e17e66b0ed76 - Cache
Resolved version ranges
    cmake/[>=3.18.0 <4]: cmake/3.31.6
    zlib/[>=1.2.11 <2]: zlib/1.3.1

======== Computing necessary packages ========
Connecting to remote 'conancenter' anonymously
Requirements
    zlib/1.3.1#b8bc2603263cf7eccbd6e17e66b0ed76:6980ffdedf2f062ab47388192e39ae3988cb5678 - Build
Build requirements
    cmake/3.31.6#ed0e6c1d49bd564ce6fed1a19653b86d:63fead0844576fc02943e16909f08fcdddd6f44b#6ec4ca55f1a1a1a6716d3c898d6624f1 - Download (conancenter)
    mold/2.32.1#e31d60209daa5ece4edf21dd3e7c15f5:c538e338200870342687a5d15d0f414fd19f440f - Build
    xxhash/0.8.2#7856c968c985b2981b707ee8f2413b2b:fa69bc2d54c599d405151e67a6a134586c0c21a4 - Build
    zlib/1.3.1#b8bc2603263cf7eccbd6e17e66b0ed76:6980ffdedf2f062ab47388192e39ae3988cb5678 - Build
ERROR: There is a loop in the graph (some packages already ommitted):
zlib/1.3.1 (Build) -> ['mold/2.32.1']
mold/2.32.1 (Build) -> ['zlib/1.3.1', 'xxhash/0.8.2', 'cmake/3.31.6']

This graph is ill-formed, and cannot be installed
Most common cause is having dependencies both in build and host contexts
forming a cycle, due to tool_requires having transitive requires.
Check your profile [tool_requires] and recipe tool and regular requires
You might inspect the dependency graph with:
  $ conan graph info . --format=html > graph.html
```

</details>

There is no real reason why this should cause a conflict: the building of a static library in the _build_ context should not be affected by any settings in the _host_ context.
Case in point, building `mold` separately first magically resolves the issue, even though this does not change the dependency graph, and this does not change any options in the build context.  

The problem here seems to be confusion between `zlib` in the host context, and `zlib` in the build context. These are independent packages, but it seems that Conan is applying the `mold` tool\_requires to both variants, even though it should _not_ apply it to `zlib` in the build context, because there is no such tool\_requires in the `build` profile.

<sub>Side note: the suggestion in the error message to use `conan graph info` to inspect the graph does not actually work, because this also crashes with the same error before producing a graph.</sub>
